### PR TITLE
Add TensorBoard support

### LINKED ISSUE #1070: Add support in selected attacks for metrics and statistics collection with TensorBoard
We plan to add support to selected attacks for collecting important metrics and statistics using TensorBoard.

Current plans include:

- option to collect all metrics/statistics every N iterations
- `ProjectedGradientDescent*` : mean loss per batch , mean L2-norm of loss gradients per batch
- `AdversarialPatch*` : adversarial patch
- for object detection estimators: mean loss per patch of each of the attack losses, mean loss per batch of the total loss

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request adds support for metrics and statistics collection with TensorBoard to selected attacks and estimators.

Fixes #1070

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py [hunk 1]
```diff
@@ -115,6 +122,7 @@ def __init__(
         self.batch_size = batch_size
         self.random_eps = random_eps
         self.verbose = verbose
+        self.tensor_board = tensor_board
```
Comment (minhitbk) [inline art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:125]: Are there any reason why not to do this in the super().__init__ function?

**Review by minhitbk — COMMENTED** [review]
Look goods.  Just one minor comment and another small question: are there any standard as to what will be necessary for an attack/estimator/... to be logged by the tensorboard?

File: art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py (hunk 1 above)
Comment (beat-buesser) [reply art/attacks/evasion/projected_gradient_descent/projected_gradient_descent.py:125]: You are right.

**Comment by beat-buesser** [comment]
No, at the moment there are no standards defined to keep it open and  flexible, but is a future release we can summarise the existing writers and  increase code reuse.

**Comment by beat-buesser** [comment]
@minhitbk Thank you for your review!

**Pull request merged**